src/dependencies.rs

```diff
@@ -0,0 +1,305 @@
+use crate::error::Error;
+pub struct FrozenDependency {
+    pub name: String,
+    pub version: String,
+}
+
+impl FrozenDependency {
+    pub fn from_string(string: &str) -> Result<Self, Error> {
+        let err = Error::BrokenPipFreezeLine {
+            line: string.to_string(),
+        };
+
+        if !string.contains("==") {
+            return Err(err);
+        }
+
+        let words: Vec<_> = string.split("==").collect();
+        if words.len() != 2 {
+            return Err(err);
+        }
+        Ok(FrozenDependency {
+            name: words[0].to_string(),
+            version: words[1].to_string(),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ParseError {
+    pub details: String,
+}
+
+impl ParseError {
+    pub fn new(details: &str) -> Self {
+        ParseError {
+            details: details.to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct VersionSpec {
+    start: usize,
+    end: usize,
+    pub value: String,
+}
+
+#[derive(Debug)]
+pub struct GitDependency {
+    pub name: String,
+    pub line: String,
+    pub git_ref: VersionSpec,
+}
+
+impl GitDependency {
+    pub fn bump(&mut self, new_ref: &str) -> bool {
+        let VersionSpec { start, end, value } = &self.git_ref;
+        if new_ref == value {
+            return false;
+        }
+        self.line = format!("{}{}{}", &self.line[0..*start], new_ref, &self.line[*end..],);
+        self.git_ref.value = new_ref.to_string();
+        true
+    }
+}
+
+#[derive(Debug)]
+pub struct SimpleDependency {
+    pub name: String,
+    pub version: VersionSpec,
+    pub line: String,
+}
+
+impl SimpleDependency {
+    pub fn from_frozen(frozen: &FrozenDependency) -> Self {
+        let name = &frozen.name;
+        let line = format!("{}=={}", name, frozen.version);
+        let version = LockedDependency::parse_simple_version(&line);
+        SimpleDependency {
+            name: name.to_string(),
```

Comment:
Alternatively, you can clone(). Usually I use to_string() to convert a `&str` to `String`, but this is a `&String`.
Not very important tho.